src/script/proof.h

```diff
@@ -0,0 +1,266 @@
+        if (m_version > BIP322_FORMAT_VERSION) throw serialization_error("Unknown BIP322 proof format; you may need to update your node");
+        if (m_version < BIP322_FORMAT_VERSION) throw serialization_error("Outdated BIP322 proof format; ask prover to re-sign using newer software");
+        READWRITE(m_entries);
+    }
+};
+
+struct SignatureProof {
+    CScript m_scriptsig; //!< ScriptSig data
+    CScriptWitness m_witness;   //!< Witness
+
+    explicit SignatureProof(const SignatureData& sigdata = SignatureData()) {
+        m_scriptsig = sigdata.scriptSig;
+        m_witness = sigdata.scriptWitness;
+    }
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(m_scriptsig);
+        READWRITE(m_witness.stack);
+    }
+};
+
+struct Purpose {
+    template<typename T>
+    Result Prepare(const T& input, std::set<T>& inputs_out) const {
+        if (inputs_out.count(input)) return RESULT_ERROR;
+        inputs_out.insert(input);
+        return RESULT_VALID;
+    }
+};
+
+/**
+ * Purpose: SignMessage
+ *
+ * Generate a sighash based on a scriptPubKey and a message. Emits VALID on success.
+ */
+struct SignMessage: Purpose {
+    CScript m_scriptpubkey;
+
+    explicit SignMessage(const CScript& scriptpubkey = CScript()) : m_scriptpubkey(scriptpubkey) {}
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(m_scriptpubkey);
+    }
+
+    Result Prepare(const std::vector<SignMessage>& entries, const std::string& message, std::set<CScript>& inputs_out, uint256& sighash_out, CScript& spk_out) const;
+
+    inline std::set<CScript> InputsSet() { return std::set<CScript>(); }
+};
+
+struct Proof: public Header {
+    std::vector<SignatureProof> m_items;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        Header::SerializationOp(s, ser_action);
+        m_items.resize(m_entries);
+        for (auto& e : m_items) {
+            READWRITE(e);
+        }
+    }
+};
+
+template<typename T>
+struct BaseWorkSpace {
+    std::map<CTxDestination, CKey> privkeys;
+    std::vector<T> m_challenge;
+    Proof m_proof;
+
+    virtual void GenerateSingleProof(const T& challenge, SigningProvider* sp, const uint256& sighash, const CScript& scriptPubKey, const std::string& message) = 0;
+    virtual Result VerifySingleProof(unsigned int flags, const T& challenge, const SignatureProof& proof, const std::string& message, const uint256& sighash, const CScript& scriptPubKey) = 0;
+
+    void Prove(const std::string& message, SigningProvider* signingProvider = nullptr) {
+        m_proof.m_items.clear();
+        m_proof.m_version = BIP322_FORMAT_VERSION;
+        m_proof.m_entries = m_challenge.size();
```

Comment:
Is `m_challenge.size() <= 255` guaranteed here (`m_entries` is `uint8_t`)? If so, perhaps `assert(...)` that to make the tacit assumption explicit? :-)

Reply:
Good point. Adding assert.